src/utils/schemaTree.js

```diff
@@ -0,0 +1,312 @@
+   */
+  if ('contains' in schema) {
+    const subschema = schema.contains;
+    const childIndex = rootNode.children.length;
+
+    subschema.contains = true;
+    createChildNode(rootNode, subschema, childIndex);
+  }
+}
+
+/**
+ * Create a tree for object data type schemas.
+ * Object properties should be created as children nodes by calling
+ * back on the createSchemaTree method and appended to given root node.
+ */
+export function createObjectTree(rootNode) {
+  const { schema } = rootNode;
+
+  /**
+   * Create child nodes only if object properties are defined.
+   */
+  if ('properties' in schema) {
+    /**
+     * Memoize required properties in advance for checking if a certain
+     * property is required within the object type schema.
+     */
+    const requiredList =
+      'required' in schema ? new Set(schema.required) : new Set();
+    const propertyList = Object.keys(schema.properties);
+
+    /**
+     * Create object properties as child nodes in sequential order.
+     * Make sure to add a name and required field for each of the subschemas
+     * so that they are displayed accordingly when creating schemaTable rows.
+     */
+    propertyList.forEach((property, childIndex) => {
+      const subschema = schema.properties[property];
+
+      subschema.name = property;
+
+      if (requiredList.has(property)) {
+        subschema.required = true;
+      }
+
+      createChildNode(rootNode, subschema, childIndex);
+    });
+  }
+}
+
+/**
+ * Fetch the reference schema defined by the refString
+ * @param {*} refString: value of $ref field within a schema
+ */
+function fetchRefSchema(schemaTree, refString) {
+  const [source, definitionPath] = refString.split('#');
+
+  /**
+   * Find the source for the reference
+   * TODO: add feature to fetch $refs in separate files or urls
+   */
+  if (source.length > 0) {
+  }
+
+  /** Find the definition within the source */
+  const parameters = definitionPath.split('/');
+  let ptr = schemaTree.schema;
+
+  parameters.forEach(parameter => {
+    if (parameter.length > 0) {
+      ptr = ptr[parameter];
+    }
+  });
+
+  return ptr;
+}
+
+/**
+ * Update the refNode's state to expanded form.
+ * @param {*} refDefaultNode: a refNode's defaultNode field
+ */
+export function expandRefNode(schemaTree, refDefaultNode) {
+  /**
+   * Create a clone of the schemaTree to maintain immutability.
+   */
+  const cloneTree = clone(schemaTree);
+
+  /**
+   * Traverse the clone tree using the refDefaultNode's path
+   * to find the corresponding ref node within the clone tree.
+   * ('nodePtr' ultimately points to a refNode with a structure
+   *   of defaultNode, expandedNode, isExpanded fields)
+   */
+  let nodePtr = cloneTree;
+
+  refDefaultNode.path.forEach(childIndex => {
+    nodePtr = nodePtr.children[childIndex];
+  });
+
+  /**
+   * Update the 'isExpanded' state of the ref node so that
+   * the ref row will now display an expanded version instead.
+   */

```

Comment:
Created a clone of the schemaTree and mutated the refNode within that clone tree, instead of using `ramda.assocPath`(https://ramdajs.com/docs/#assocPath) since using the latter would lead to creating some unnecessary changes in the node's structure.

Specifically, using `assocPath` would lead to altering the structure of certain nodes' `children` fields from its initial array format to an object format. That is, the nodes that are on the path from the root to the concerning refNode will have the `children` fields changed to an object format, while the rest nodes will have `children` fields as an array format. 
```js
// original format of node         // mutation on certain nodes due to assocPath
{                                  {
  schema: { .. },                     // .. same
  path: [ .. ],                       
  children: [                         children: {
    { node object },                     0: { node object },
    { node object },                     1: { node object },
  ],                                  },
}                                   }
```

This is mainly because the `assocPath`'s path argument cannot accept array notation but only object key values. For example, if the concerning refNode can be accessed via `children[0] - children[1] - children[2]` (the root node -> 0th index child node ->1st index child node ->  2nd index child node), assocPath's path argument must be `['children', '0', 'children', '1', 'children', '2']` in order to access the refNode.  (`['children[0', 'children[1]', 'children[2]']` does not work)
And as a consequence of using that as a path parameter, some node structures are changed.

I do think that we can make use of `assocPath` if we add some code to deal when a node's `children` property is an object type by changing it to an array format every time we need to access the `children` property, but it felt a bit messy in that some node's have an array structure and some have an object structure.

Which way do you think would be better: using the `clone` so that we can maintain the same structure for all nodes, or using the `assocPath` and adding some extra code to deal with different structures?

Reply:
`clone` should be okay.